lib/graphql/graphql-provider.ts

```diff
@@ -0,0 +1,61 @@
+import { ChainId } from '@uniswap/sdk-core'
+
+import { GraphQLClient, IGraphQLClient } from './graphql-client'
+import {
+  GRAPHQL_QUERY_MULTIPLE_TOKEN_INFO_BY_CONTRACTS,
+  GRAPHQL_QUERY_TOKEN_INFO_BY_ADDRESS_CHAIN,
+} from './graphql-queries'
+import { TokenInfoResponse, TokensInfoResponse } from './graphql-schemas'
+
+/* Interface for accessing Uniswap GraphQL API */
+export interface IUniGraphQLProvider {
+  /* Fetch token info for a given chain and address */
+  getTokenInfo(chainId: ChainId, address: string): Promise<TokenInfoResponse>
+  /* Fetch token info for multiple tokens given a chain and addresses */
+  getTokensInfo(chainId: ChainId, addresses: string[]): Promise<TokensInfoResponse>
+  // Add more methods here as needed.
+  // - more details: https://github.com/Uniswap/data-api-graphql/blob/main/graphql/schema.graphql
+}
+
+/* Implementation of the UniGraphQLProvider interface to give access to Uniswap GraphQL API */
+export class UniGraphQLProvider implements IUniGraphQLProvider {
+  private readonly endpoint = process.env.UNI_GRAPHQL_ENDPOINT!
+  private readonly headers = {
+    Origin: process.env.UNI_GRAPHQL_HEADER_ORIGIN!,
+    'Content-Type': 'application/json',
+  }
+  private client: IGraphQLClient
+
+  constructor() {
+    this.client = new GraphQLClient(this.endpoint, this.headers)
+  }
+
+  /* Convert ChainId to a string recognized by data-graph-api graphql endpoint.
+   *  GraphQL Chain Enum located here: https://github.com/Uniswap/data-api-graphql/blob/main/graphql/schema.graphql#L155
+   *  */
+  private _chainIdToGraphQLChainName(chainId: ChainId): string | undefined {
```

Comment:
I understand that GQL only supports mainnet for FOT fee fetching, but I think from routing-api perspective, we probably don't have to code this logic ourselves, because:
1. it's easy to forget that we also have a per-chain gatekeeping
2. on the top-level token fee fetcher that we plan to implement, we can also another in-memory cache layer, so that the first time we hit GQL and get error due to chain not supported, we can fallback to SOR and cache the result for the entire duration of the lambda cycle. in other words, as long as we implement the top-level token fee fetcher right, we won't DDOS against GQL endpoint on the unsupported chains.

Reply:
for # 2, this situation would be handled by our onchain fallback.
for # 1, I agree but this method here is necessary as we want to translate [sdk-core/ChainId enum](https://github.com/Uniswap/sdk-core/blob/main/src/chains.ts#L1) to a string that matches [data-api-graphql/Chain enum](https://github.com/Uniswap/data-api-graphql/blob/main/graphql/schema.graphql#L155). Unfortunately diff enums are used in different places in our repos. We should probably ask data team to provide (and maintain) this conversion method for us eventually.